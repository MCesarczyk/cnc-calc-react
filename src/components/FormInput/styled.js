import styled from "styled-components";

export const InputLabelText = styled.span`
    text-align: right;

    @media (max-width: 480px) {
        text-align: center;
    }
`;