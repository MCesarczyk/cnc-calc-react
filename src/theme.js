export const theme = {
    color: {
        primary: "#1e3246",
        background: "#133246",
    },
    breakpoint: {
        mobileMax: "480px",
    },
}