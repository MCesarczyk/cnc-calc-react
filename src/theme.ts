export const theme = {
  color: {
    primary: '#1e3246',
    primaryLight: '#2e4c6b',
    primaryText: '#000',
    contrastText: '#fff',
    background: '#133246',
    elementBackground: '#fff',
    submitButtonText: '#19aa6e',
    resetButtonText: '#dc3223',
    copyButtonText: '#dc143c',
    activeNavlinkText: '#2188FF',
    invalid: 'beige',
    invalidBackground: 'beige',
    disabled: '#ddd',
    disabledText: '#808080',
    disabledBackground: '#ddd',
    placeholder: '#747474',
  },
  breakpoint: {
    xs: '480px',
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
    xxl: '1400px',
  },
};
