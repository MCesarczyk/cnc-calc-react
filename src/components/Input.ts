import styled from 'styled-components';

export const Input = styled.input`
  border: solid 1px ${({ theme }) => theme.color.primary};
  border-radius: 5px;
  padding: 5px;

  @media (max-width: ${({ theme }) => theme.breakpoint.xs}) {
    width: 100%;
  }

  &:invalid {
    background-color: ${({ theme }) => theme.color.invalid};
  }

  &:disabled {
    background-color: ${({ theme }) => theme.color.disabled};
  }
`;
