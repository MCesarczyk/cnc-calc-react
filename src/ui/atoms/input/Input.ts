import { CSSProperties } from 'react';
import styled from 'styled-components';

export type InputVariants = 'DEFAULT';

type InputStyles = Required<
  Pick<CSSProperties, 'color' | 'backgroundColor' | 'borderColor'> & {
    invalidBackgroundColor?: CSSProperties['backgroundColor'];
    disabledBackgroundColor?: CSSProperties['backgroundColor'];
  }
>;

export type Palette<TVariants extends string, TStyles> = {
  [variant in TVariants]: TStyles;
};

const styles: Palette<InputVariants, InputStyles> = {
  DEFAULT: {
    color: 'primaryText',
    backgroundColor: 'elementBackground',
    borderColor: 'primary',
    invalidBackgroundColor: 'invalid',
    disabledBackgroundColor: 'disabledBackground',
  },
};

interface InputProps {
  $variant?: InputVariants;
  placeholder?: string;
  required?: boolean;
  disabled?: boolean;
}

export const Input = styled.input<InputProps>`
  border: solid 1px ${(props) => props.theme.color[styles[props.$variant!].borderColor]};
  border-radius: 5px;
  padding: 5px;

  @media (max-width: ${(props) => props.theme.breakpoint.xs}) {
    width: 100%;
  }

  &:invalid {
    background-color: ${(props) => props.theme.color[styles[props.$variant!].invalidBackgroundColor]};
  }

  &:disabled {
    background-color: ${(props) => props.theme.color[styles[props.$variant!].disabledBackgroundColor]};
  }
`;

Input.defaultProps = {
  $variant: 'DEFAULT',
};

Input.displayName = 'Input';
