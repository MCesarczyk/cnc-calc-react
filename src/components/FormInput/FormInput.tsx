import { FieldWrapper } from '../FieldWrapper';
import { Input } from '../Input';
import { TextSub } from '../TextSub/TextSub';
import { Label } from '../Label';
import { ChangeEvent } from 'react';

interface FormInputProps {
  name: string;
  sub?: string;
  unit?: string;
  inputRef?: any;
  value: string;
  min?: number;
  step?: number;
  placeholder?: string;
  disabled?: boolean;
  readOnly?: boolean;
  autoFocus?: boolean;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export const FormInput = ({
  name,
  sub,
  unit,
  inputRef,
  value,
  min,
  step,
  placeholder,
  disabled,
  readOnly,
  autoFocus,
  onChange,
}: FormInputProps) => {
  return (
    <FieldWrapper>
      <Label>
        <label>
          {name}
          <TextSub>{sub}</TextSub>
          {unit}
        </label>
      </Label>
      <Input
        ref={inputRef}
        value={value}
        type="number"
        min={min}
        step={step}
        placeholder={placeholder}
        required
        disabled={disabled}
        readOnly={readOnly}
        autoFocus={autoFocus}
        onChange={onChange}
      />
    </FieldWrapper>
  );
};
