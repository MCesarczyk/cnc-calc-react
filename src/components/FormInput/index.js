import { Input } from "../Input";
import { TextSub } from "../TextSub";
import { InputLabelText, FieldWrapper } from "./styled";

const FormInput = ({
  name,
  sub,
  unit,
  inputRef,
  value,
  type,
  min,
  step,
  placeholder,
  required,
  disabled,
  readOnly,
  autoFocus,
  onChange
}) => {
  return (
    <FieldWrapper>
      <InputLabelText>
        <label>
          {name}
          <TextSub>
            {sub}
          </TextSub>
          {unit}
        </label>
      </InputLabelText>
      <Input
        ref={inputRef}
        value={value}
        type={type}
        min={min}
        step={step}
        placeholder={placeholder}
        required={required}
        disabled={disabled}
        readOnly={readOnly}
        autoFocus={autoFocus}
        onChange={onChange}
      />
    </FieldWrapper>
  )
};

export default FormInput;