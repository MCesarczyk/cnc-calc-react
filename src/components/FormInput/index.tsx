import { FieldWrapper } from "../FieldWrapper";
import { Input } from "../Input";
import { TextSub } from "../TextSub";
import { Label } from "../Label";

interface FormInputProps {
  name: string;
  sub?: string;
  unit?: string;
  inputRef?: any;
  value?: any;
  min?: number;
  step?: number;
  placeholder?: string;
  disabled?: boolean;
  readOnly?: boolean;
  autoFocus?: boolean;
  onChange?: any;
}

const FormInput = ({
  name,
  sub,
  unit,
  inputRef,
  value,
  min,
  step,
  placeholder,
  disabled,
  readOnly,
  autoFocus,
  onChange
}:FormInputProps) => {
  return (
    <FieldWrapper>
      <Label>
        <label>
          {name}
          <TextSub>
            {sub}
          </TextSub>
          {unit}
        </label>
      </Label>
      <Input
        ref={inputRef}
        value={value}
        type="number"
        min={min}
        step={step}
        placeholder={placeholder}
        required
        disabled={disabled}
        readOnly={readOnly}
        autoFocus={autoFocus}
        onChange={onChange}
      />
    </FieldWrapper>
  )
};

export default FormInput;