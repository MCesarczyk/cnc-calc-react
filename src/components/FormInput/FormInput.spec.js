import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { theme } from 'theme';
import { Input } from '../Input';
import { FormInput } from './FormInput';

const setup = () =>
  render(
    <ThemeProvider theme={theme}>
      <FormInput />
    </ThemeProvider>,
  );

test('Input component should be empty initially', () => {
  const input = setup().getByRole('spinbutton');
  expect(input.value).toBe('');
});

test('Input component should display changed value of state variable', () => {
  const input = setup().getByRole('spinbutton');
  fireEvent.change(input, { target: { value: '123' } });
  expect(input.value).toBe('123');
});

test("Input component shouldn't allow to input letters", () => {
  const input = setup().getByRole('spinbutton');
  fireEvent.change(input, { target: { value: 'Qwerty' } });
  expect(input.value).toBe('');
});

test('FormInput component rendering properly', () => {
  const input = renderer.create(
    <ThemeProvider theme={theme}>
      <FormInput />
    </ThemeProvider>,
  );

  let tree = input.toJSON();

  expect(tree).toMatchSnapshot();
});

test('Input component change style on actions', () => {
  const input = renderer.create(
    <ThemeProvider theme={theme}>
      <Input />
    </ThemeProvider>,
  );

  let tree = input.toJSON();

  expect(tree).toHaveStyleRule('background-color', 'beige', {
    modifier: ':invalid',
  });

  expect(tree).toHaveStyleRule('background-color', '#ddd', {
    modifier: ':disabled',
  });
});
