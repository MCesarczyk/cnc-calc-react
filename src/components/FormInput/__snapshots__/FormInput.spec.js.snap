// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component rendering properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  min-width: 16rem;
  border: solid 1px #1e3246;
  border-radius: 5px;
  padding: 5px;
}

.c2:invalid {
  background-color: beige;
}

.c2:disabled {
  background-color: #ddd;
}

.c1 {
  text-align: right;
}

@media (max-width:480px) {
  .c0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
    gap: 0.25rem;
  }
}

@media (max-width:480px) {
  .c2 {
    width: 100%;
  }
}

@media (max-width:480px) {
  .c1 {
    -webkit-align-self: flex-start;
    -ms-flex-item-align: start;
    align-self: flex-start;
  }
}

<div
  className="c0"
>
  <label
    className="c1"
  />
  <input
    className="c2"
  />
</div>
`;
